eigensync/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+//use crate::protocol::Request;
+use automerge::{ActorId, AutoCommit, Change};
+use autosurgeon::{hydrate, reconcile, Hydrate, Reconcile};
+use tokio_util::task::AbortOnDropHandle;
+use crate::protocol::{client, Behaviour, BehaviourEvent, ChannelRequest, Response, SerializedChange, SignedEncryptedSerializedChange, ServerRequest};
+
+pub type SyncBehaviour = request_response::cbor::Behaviour<ServerRequest, Response>;
+
+pub struct SyncLoop {
+    receiver: UnboundedReceiver<ChannelRequest>,
+    response_map: HashMap<OutboundRequestId, oneshot::Sender<Result<Vec<SignedEncryptedSerializedChange>, String>>>,
+    swarm: Swarm<Behaviour>,
+    connection_established: Option<oneshot::Sender<()>>,
+}
+
+impl SyncLoop {
+    pub async fn new(receiver: UnboundedReceiver<ChannelRequest>, swarm: Swarm<Behaviour>, connection_established: oneshot::Sender<()>) -> Self {
+        
+        Self { receiver, response_map: HashMap::new(), swarm, connection_established: Some(connection_established)}
+    }
+    
+    pub fn sync_with_server(&mut self, request: ChannelRequest, server_id: PeerId) {
+        let server_request = ServerRequest::UploadChangesToServer { encrypted_changes: request.encrypted_changes };
+        match &server_request {
+            ServerRequest::UploadChangesToServer { encrypted_changes: _changes } => {
+                let request_id = self.swarm.behaviour_mut().send_request(&server_id, server_request);
+                self.response_map.insert(request_id, request.response_channel);
+            },
+        }
+    }
+    
+    pub async fn run(&mut self, server_id: PeerId) -> anyhow::Result<()> {
+        loop {
+            tokio::select! {
+                event = self.swarm.select_next_some() => handle_event(event, server_id, &mut self.swarm, &mut self.response_map, self.connection_established.take()).await?,
+                request_from_handle = self.receiver.recv() => {
+                    if let Some(request) = request_from_handle {
+                        self.sync_with_server(request, server_id);
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub async fn handle_event(
+    event: SwarmEvent<BehaviourEvent>,
+    server_id: PeerId,
+    swarm: &mut Swarm<Behaviour>,
+    response_map: &mut HashMap<OutboundRequestId, oneshot::Sender<Result<Vec<SignedEncryptedSerializedChange>, String>>>,
+    mut connection_established: Option<oneshot::Sender<()>>,
+) -> anyhow::Result<()> {
+    Ok(match event {
+        SwarmEvent::Behaviour(BehaviourEvent::Sync(request_response::Event::Message {
+            peer,
+            message,
+        })) => match message {
+            request_response::Message::Response {
+                request_id,
+                response,
+            } => match response {
+                Response::NewChanges { changes } => {
+                    let sender = match response_map.remove(&request_id) {
```

Comment:
I.e., this match statement can be replaced with `.context(format!("No sender for request id {:?}", request_id))?`